### ISSUE TITLE
Feature Request:Moloch unable to classify single-packet HTTP traffic 

### ISSUE THREAD

**Issue by bokandbok** [aol/moloch#880]
The current latest version of Moloch version 1.1.1 does not classify the protocol if the traffic is a single-packet HTTP traffic. I have checked with Andy and he shared that it is an expected behaviour as Moloch is designed to classify sessionised traffic. 

I think we will have many use-cases where we only have a single-packet traffic (E.g Application-layer HTTP). Will like to see Moloch supporting that. For your considerations. Thanks :)

**Comment by bokandbok** [comment]
Hi Andy, i have opened an issue "Feature Request:Moloch unable to classify single-packet HTTP traffic" a couple of days ago. No one commented. Does that mean the feature request will be worked on or not? Thanks.

**Comment by awick** [comment]
We usually don't comment if we are working on a feature or not.  I think it would be useful if you provide some same pcaps and exactly what you expect to happen around all the many edge cases.

**Comment by bokandbok** [comment]
I see. actually I am not too sure if the pcaps are useful. It is just be a collection of layer-7 HTTP packets without all the usual TCP streams (E.g SYN, ACK etc). I do hope that Moloch can support classifying the HTTP packets even without the streams. Thanks

**Comment by awick** [comment]
* In general we don't implement new features/protocols without sample pcap.  You providing that pcap (that can be publicly added to the test suite) would definitely reduce the friction of a feature being implemented.  I'm not clear why you think it wouldn't be useful.

* So you only want this feature for single packet streams?  So if there are two packets it wouldn't work?

* How do we know this is a single packet stream?  Just wait for all the timeouts to expire?

**Comment by bokandbok** [comment]
Noted. i have attached a sample PCAP here. In summary, I will like to raise 2 points here:

1)  Moloch does not attempt to classify these packets as HTTP under "SPI View" tab
2) Another issue here is Under "Session" tab, there seem to have an issue when exporting out sessions originated from pcapng format (The pcap filesize is always 0 bytes). This problem does not exist if 
my packets are in pcap format.

Thanks andy for your kind advice.

[sample.zip](https://github.com/aol/moloch/files/2227973/sample.zip)

**Comment by awick** [comment]
You didn't answer the questions above.  note, most of these sessions have more then 1 packet.  

> Moloch does not attempt to classify these packets as HTTP under "SPI View" tab

The tab doesn't matter, as stated before moloch only classifies sessions

> pcapng

pcapng isn't well supported in moloch yet, feel free to open another issues (although I think there might be one already)

**Comment by bokandbok** [comment]
sorry andy. Forgotten to answer your earlier Qs.

I have resubmitted another sample showing a more accurate scenario.
[sample2.zip](https://github.com/aol/moloch/files/2228096/sample2.zip)

To answer your earlier Qs,
In general we don't implement new features/protocols without sample pcap. You providing that pcap (that can be publicly added to the test suite) would definitely reduce the friction of a feature being implemented. I'm not clear why you think it wouldn't be useful.
**I have resubmitted a more accurate sample. Please take a look.**

So you only want this feature for single packet streams? So if there are two packets it wouldn't work?
**I mean i will still want Moloch to work on classifying HTTP traffic with multiple packets streams, just hope to include an additional support for single-packet stream (similar to my 2nd sample).**


How do we know this is a single packet stream? Just wait for all the timeouts to expire?
**I ingested the pcaps into Moloch, not through capturing the packets through network interface.
I guess if the source is originated from pcap, we can tell immediately if the stream is a single packet?
As what you have advised, can we consider using the timeout if the traffic source is originated from network interface?**

Thanks andy

**Comment by awick** [comment]
>>So you only want this feature for single packet streams? So if there are two packets it wouldn't work?

>I mean i will still want Moloch to work on classifying HTTP traffic with multiple packets streams, just hope to include an additional support for single-packet stream (similar to my 2nd sample).

So to answer the question, you only want this feature to work on single packet streams?


>>How do we know this is a single packet stream? Just wait for all the timeouts to expire?

> I ingested the pcaps into Moloch, not through capturing the packets through network interface.
> I guess if the source is originated from pcap, we can tell immediately if the stream is a single packet?

You can't tell immediately, but you could tell at the end of the file I guess.  Might as well use the timeouts.  (The timeouts auto fire when capture exists or --flush is used)

> As what you have advised, can we consider using the timeout if the traffic source is originated from network interface?


So to summarize.  If a tcp session only has 1 packet and once all the timeouts expire you want to run the classify/parsers against the single packet?

**Comment by bokandbok** [comment]
So you only want this feature for single packet streams? So if there are two packets it wouldn't work?

I mean i will still want Moloch to work on classifying HTTP traffic with multiple packets streams, just hope to include an additional support for single-packet stream (similar to my 2nd sample).

So to answer the question, you only want this feature to work on single packet streams?
**Yes. You are right**

How do we know this is a single packet stream? Just wait for all the timeouts to expire?

I ingested the pcaps into Moloch, not through capturing the packets through network interface.
I guess if the source is originated from pcap, we can tell immediately if the stream is a single packet?

You can't tell immediately, but you could tell at the end of the file I guess. Might as well use the timeouts. (The timeouts auto fire when capture exists or --flush is used)

As what you have advised, can we consider using the timeout if the traffic source is originated from network interface?

So to summarize. If a tcp session only has 1 packet and once all the timeouts expire you want to run the classify/parsers against the single packet?
**Yes. Agreed that timeout is a better indicator to determine whether it is a single-packet stream regardless o whether the source is a pcap or network interface.**

Thanks
